API/BUG: DatetimeIndex correctly localizes integer data

mroeschke opened the pull request:
closes #20997 
closes #20964
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I did a little refactoring of `Datetime.__new__` that converts passed data but made it easier to fix this issue. I had to modify some methods (join, normalize, delete, insert) and tests that weren't assuming integer data weren't epoch timestamps.

mroeschke committed:
BUG: Localize Index with datetime dtype and integer data correctly

mroeschke committed:
Adjust tests

mroeschke committed:
Merge remote-tracking branch 'upstream/master' into index_datetime_int

mroeschke committed:
Adjust tests

mroeschke committed:
Merge remote-tracking branch 'upstream/master' into index_datetime_int

mroeschke committed:
Refactor processing logic

mroeschke committed:
Merge remote-tracking branch 'upstream/master' into index_datetime_int_redux

mroeschke committed:
Adjust test

mroeschke committed:
Merge remote-tracking branch 'upstream/master' into index_datetime_int_redux

mroeschke committed:
add additional comment

mroeschke committed:
Merge remote-tracking branch 'upstream/master' into index_datetime_int_redux

mroeschke committed:
adjust blank spaces

jreback commented on pandas/tests/indexes/datetimes/test_construction.py:
```diff
@@ -33,16 +33,15 @@ def test_construction_with_alt(self):
         tm.assert_index_equal(i, i2)
         assert i.tz.zone == 'US/Eastern'
 
-        i2 = DatetimeIndex(i.tz_localize(None).asi8, dtype=i.dtype)
```
i actually like this construction, can you add a test that does both? (e.g. both uses and does not localize), maybe split it out and parameterize

jreback commented on pandas/tests/indexes/datetimes/test_construction.py:
```diff
@@ -469,6 +468,14 @@ def test_constructor_with_non_normalized_pytz(self, tz):
+    @pytest.mark.parametrize('klass', [Index, DatetimeIndex])
+    @pytest.mark.parametrize('box', [np.array, list])
+    def test_constructor_with_int_tz(self, klass, box):
+        ts = Timestamp('2018-01-01', tz='US/Pacific')
```
can you add the issue number

jreback commented on pandas/tests/indexes/test_base.py:
```diff
@@ -402,26 +402,30 @@ def test_constructor_dtypes_to_timedelta(self, cast_index, vals):
     @pytest.mark.parametrize("klass", [pd.Index, pd.DatetimeIndex])
-    def test_constructor_dtypes_datetime(self, tz_naive_fixture, values,
+    def test_constructor_dtypes_datetime(self, tz_naive_fixture, attr, utc,
                                          klass):
```
can you add a bit of commentary on what is tested here

jreback reviewed (changes requested):
ok looks pretty good. 

- can you add a whatsnew note (0.24.0)
- can you test with .astype as well
- can you search thru some issue for timezone / DTI. IIRC I think this might solve some (just a guess)

mroeschke committed:
Merge remote-tracking branch 'upstream/master' into index_datetime_int_redux

mroeschke committed:
Address review

mroeschke committed:
Merge remote-tracking branch 'upstream/master' into index_datetime_int_redux

mroeschke committed:
Adjust astype

mroeschke commented:
Thanks @jreback. Moved the whatsnew to v0.24, added some additional logic to get `astype` working. It doesn't appear to fix any of the other reported timezone issues.

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -806,8 +792,9 @@ def _mpl_repr(self):
+        """Return a boolean if we are only dates (and don't have a timezone)"""
         from pandas.io.formats.format import _is_dates_only
-        return _is_dates_only(self.values)
+        return _is_dates_only(self.values) and self.tz is None
```
why did this need changing?

mroeschke replied:
After this refactor, I was getting errors where the repr didn't look correct but the i8 data was correct in the case when a tz was passed in the `DatetimeIndex` constructor.

`_is_dates_only` tries to:
> return a boolean if we are only dates (and don't have a timezone)

but checks for timezones by re-passing `self.values` into `DatetimeIndex`. I am not sure the historical reason why this was the case, but checking `self.tz is None` fixed the repr issue (and seems to achieve what that function was trying to do in the first place)

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -394,57 +394,43 @@ def __new__(cls, data=None,
-                raise TypeError('cannot convert Int64Index->DatetimeIndex')
+            # assume this data are epoch timestamps
             if data.dtype != _INT64_DTYPE:
                 data = data.astype(np.int64)
```
could
``data = data.astype(np.int64, copy=False)``

jreback replied:
I don't think we have any tests with a non-int64 dtype, can you add something

jreback commented on pandas/tests/indexes/datetimes/test_astype.py:
```diff
@@ -225,6 +225,13 @@ def _check_rng(rng):
         _check_rng(rng_eastern)
         _check_rng(rng_utc)
 
+    def test_integer_index_astype_datetimetz_dtype(self):
```
can you use tz_naive_fixture here

jreback commented on pandas/tests/indexes/datetimes/test_construction.py:
```diff
@@ -26,25 +27,28 @@ def test_construction_caching(self):
 
-    def test_construction_with_alt(self):
-
+    @pytest.mark.parametrize('kwargs', [
```
can you parameterize on tz_aware_fixture (doesn't include None), then parameterize as you are doing

jreback commented on pandas/tests/indexes/datetimes/test_construction.py:
```diff
@@ -26,25 +27,28 @@ def test_construction_caching(self):
+        tm.assert_index_equal(i, result)
+        assert result.tz.zone == 'US/Eastern'
+
+    @pytest.mark.parametrize('kwargs', [
```
same

jreback commented on pandas/tests/indexes/datetimes/test_construction.py:
```diff
@@ -26,25 +27,28 @@ def test_construction_caching(self):
+    @pytest.mark.parametrize('kwargs', [
+        {'tz': 'dtype.tz'},
+        {'dtype': 'dtype'},
+        {'dtype': 'dtype', 'tz': 'dtype.tz'}])
```
can you explain how this test is different than the one above?

mroeschke replied:
I am preserving the construction of `tz_localize(None)` that was here before but split it out in its own test as discussed here https://github.com/pandas-dev/pandas/pull/21216#discussion_r191286077

jreback commented on pandas/tests/indexes/datetimes/test_construction.py:
```diff
@@ -469,6 +473,15 @@ def test_constructor_with_non_normalized_pytz(self, tz):
 
+    @pytest.mark.parametrize('klass', [Index, DatetimeIndex])
+    @pytest.mark.parametrize('box', [np.array, list])
+    def test_constructor_with_int_tz(self, klass, box):
```
parameterize on tz_naive_fixture

mroeschke committed:
Merge remote-tracking branch 'upstream/master' into index_datetime_int_redux

mroeschke committed:
Add fixture tests, test with object data, dont copy with astype

jreback commented:
looks good. @jschendel @WillAyd if you can have a look if any comments.

mroeschke committed:
Merge remote-tracking branch 'upstream/master' into index_datetime_int_redux

mroeschke committed:
Merge remote-tracking branch 'upstream/master' into index_datetime_int_redux

mroeschke committed:
add assertion back after merge conflict

jreback commented:
thanks @mroeschke

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -1175,6 +1175,10 @@ def astype(self, dtype, copy=True):
                                     copy=copy)
         try:
+            if is_datetime64tz_dtype(dtype):
+                from pandas.core.indexes.datetimes import DatetimeIndex
```
side issue (another PR). The imports inside functions should be as simple as possible, e.g.
from pandas import DatetimeIndex (several occurrences of this)
